src/optimizer/statistics/stats_storage.cpp

```diff
@@ -6,39 +6,86 @@
 #include "loggers/optimizer_logger.h"
 
 namespace noisepage::optimizer {
-common::ManagedPointer<TableStats> StatsStorage::GetTableStats(catalog::db_oid_t database_id,
-                                                               catalog::table_oid_t table_id) {
-  StatsStorageKey stats_storage_key = std::make_pair(database_id, table_id);
+
+/*
+ * Currently when getting statistics for a table we get and cache the statistics for the entire table. If we find that
+ * large tables are filling up the cache when we only need some of the columns, we may want to revisit this so we only
+ * cache the columns we use.
+ */
+StatsStorageReference StatsStorage::GetTableStats(const catalog::db_oid_t database_id,
+                                                  const catalog::table_oid_t table_id,
+                                                  catalog::CatalogAccessor *accessor) {
+  if (!ContainsTableStats(database_id, table_id)) {
+    InsertTableStats(database_id, table_id, accessor);
+  }
+
+  common::SharedLatchGuard shared_stats_storage_latch{&stats_storage_latch_};
+
+  StatsStorageKey stats_storage_key{database_id, table_id};
   auto table_it = table_stats_storage_.find(stats_storage_key);
 
-  if (table_it != table_stats_storage_.end()) {
-    return common::ManagedPointer<TableStats>(table_it->second);
+  auto &stats_storage_value = table_it->second;
+
+  UpdateStaleColumns(table_id, &stats_storage_value, accessor);
+
+  return StatsStorageReference(&stats_storage_value, std::move(shared_stats_storage_latch));
+}
+
+void StatsStorage::MarkStatsStale(catalog::db_oid_t database_id, catalog::table_oid_t table_id,
+                                  const std::vector<catalog::col_oid_t> &col_ids) {
+  /*
+   * It's ok to mark something stale while someone is reading it, the worst that happens is they end up using slightly
+   * stale statistics without realizing it.
```

Comment:
But what if someone is updating it through `UpdateStaleColumns`? Wouldn't it be possible that the value is corrupted? I think you still need to take the shared latch on the `stats_storage_value`.

Reply:
Thanks for catching this (and being patient with me not fixing this correctly multiple times). I meant to take a read lock on the table stats and this comment was explaining why I take a read lock instead of a write lock. It should be fixed now.